pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -91,177 +185,196 @@ func (w *Watcher) traversePluginDir(dir string) error {
 	}
 	defer conn.Close()
 
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
+
 	infoResp, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
 	if err != nil {
 		return fmt.Errorf("failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
 	}
 
-	return w.invokeRegistrationCallbackAtHandler(ctx, client, infoResp, socketPath)
-}
-
-func (w *Watcher) invokeRegistrationCallbackAtHandler(ctx context.Context, client registerapi.RegistrationClient, infoResp *registerapi.PluginInfo, socketPath string) error {
-	var handlerCbkFn RegisterCallbackFn
-	var ok bool
-	handlerCbkFn, ok = w.handlers[infoResp.Type]
+	handler, ok := w.handlers[infoResp.Type]
 	if !ok {
-		errStr := fmt.Sprintf("no handler registered for plugin type: %s at socket %s", infoResp.Type, socketPath)
-		if _, err := client.NotifyRegistrationStatus(ctx, &registerapi.RegistrationStatus{
-			PluginRegistered: false,
-			Error:            errStr,
-		}); err != nil {
-			return errors.Wrap(err, errStr)
-		}
-		return errors.New(errStr)
+		return w.notifyPlugin(client, false, fmt.Sprintf("no handler registered for plugin type: %s at socket %s", infoResp.Type, socketPath))
 	}
 
-	var versions []string
-	for _, version := range infoResp.SupportedVersions {
-		versions = append(versions, version)
+	// ReRegistration: We want to handle multiple plugins registering at the same time with the same name sequentially.
+	// See the state machine for more information.
+	// This is done by using a Lock for each plugin with the same name and type
+	pool := w.getPluginPool(infoResp.Type, infoResp.Name)
+
+	pool.Lock()
+	defer pool.Unlock()
+
+	if infoResp.Endpoint == "" {
+		infoResp.Endpoint = socketPath
 	}
+
 	// calls handler callback to verify registration request
-	chanForAckOfNotification, err := handlerCbkFn(infoResp.Name, infoResp.Endpoint, versions, socketPath)
-	if err != nil {
-		errStr := fmt.Sprintf("plugin registration failed with err: %v", err)
-		if _, err := client.NotifyRegistrationStatus(ctx, &registerapi.RegistrationStatus{
-			PluginRegistered: false,
-			Error:            errStr,
-		}); err != nil {
-			return errors.Wrap(err, errStr)
-		}
-		return errors.New(errStr)
+	if err := handler.ValidatePlugin(infoResp.Name, infoResp.Endpoint, infoResp.SupportedVersions); err != nil {
+		return w.notifyPlugin(client, false, fmt.Sprintf("plugin registration failed with err: %v", err))
+	}
+
+	if err := handler.InitPlugin(infoResp.Name, infoResp.Endpoint); err != nil {
+		return w.notifyPlugin(client, false, fmt.Sprintf("plugin registration failed with err: %v", err))
 	}
 
-	if _, err := client.NotifyRegistrationStatus(ctx, &registerapi.RegistrationStatus{
-		PluginRegistered: true,
-	}); err != nil {
-		chanForAckOfNotification <- false
+	// pluginwatcher Register is called before plugin consumer register so that a plugin
+	// that is removed while the consumer registers will then see a deRegister call.
```

Comment:
Nope. the pluginwatcher's registerPlugins adds the plugin to the plugins map.
Once it's in the map we can make the connection socketPath -> plugin

Reply:
Reworded to the following:
```
        // We add the plugin to the pluginwatcher's map before calling a plugin consumer's Register handle
        // so that if we recieve a delete event during Register Plugin, we can process it as a DeRegister call.
```